### DISCUSSION TITLE
Enabling Inheritance on Poet Classes & Builders

### DISCUSSION THREAD

**Discussion by brizzbuzz** [square/kotlinpoet#1081] [Ideas]
Wanted to gauge community interest in allowing KotlinPoet class declarations to be extended.  By default Kotlin classes are private, so this would mostly just require adding the `open` modifier to most classes. 

The reason being, there are certain things that developers may want to build (in my case, certain DSL streamline functions) that most likely don't really belong in the KotlinPoet repo.  Opening up these classes would allow external extension that is currently not feasible. 

Curious to hear if there are any arguments _against_ allowing this feature.

**Comment by ZacSweers** [comment]
Isn't this what extension functions are for?

**Reply by brizzbuzz** [reply]
Kinda, but there are limitations.  Scoping and infix functions are the main two that I've encountered.  

Example: Let's combine the two into some pseudo Kotlin code 

`infix fun CodeBlock.Builder.String.statement(args: List<Any>)`

What this would do is, in the scope of a CodeBlock builder, allow you to have an infix function to streamline the declaration of a statement.  

The reason I would argue that these should be opened up is that niche stuff like this most likely doesn't belong in the KotlinPoet repo, but non the less, developers should have the flexibility to do it if they choose.  

But because the KotlinPoet classes are final, you can't extend it and add your own features, so you cannot achieve this as it stands.

**Reply by Egorand** [reply]
I don't really follow, would you be able to fork the repo, `open` the classes you need, implement the feature you're looking for and paste a link to the commit here, so it's more clear what the use case is?

**Reply by brizzbuzz** [reply]
Hmm would be a little annoying to show an example in KotlinPoet, so I just put the example in the repo I'm working out of for these misc DSLs I need.... here is a PR https://github.com/lg-backbone/poetik/pull/4 

Hopefully it gives enough insight into the use case.

**Reply by ZacSweers** [reply]
What you want is multiple receivers on extension functions, which is being explored here: https://youtrack.jetbrains.com/issue/KT-10468

I don't think your use case alone is a compelling enough reason to make these classes open.

**Reply by brizzbuzz** [reply]
I disagree, it's not so much one compelling use case or another, but rather that leaving them closed prevents other users from leveraging KotlinPoet in the way they choose.  

So rather than assessing this one use case, I would ask what the gain is of leaving them final.

**Reply by brizzbuzz** [reply]
Side note, I'm not holding out hope for multiple receivers any time soon it's been a tracked issue for over half a decade

**Reply by ZacSweers** [reply]
Sorry you'll need to take no for an answer here. Inheritance must be designed for, open by default is an anti-pattern and there's a reason kotlin is closed by default. The question is always why, not why not.

**Reply by Egorand** [reply]
+1, these classes haven't been designed to be open, if we just make them `open` we'll likely run into bugs we won't want to fix. And there's no going back too, so unless there's a very very compelling case to have them be open, we won't consider this. What you're running into does feel like a language limitation, and not an issue with how the library is designed.
